### PR TITLE
Docs rewrite

### PR DESCRIPTION
**Pull request by daveads**
###  Roll out Comprehensive documentation for ZIO HTTP #2198 

Would like to keep experimenting and exploring zio-http to better understand what's going on and constantly update the doc...

*  Introduction ✅ 
*  Concepts 
     * Routing ✅ 
     * Request Handling ✅ 
     * Server ✅ 
     * Client ✅ 
     * Middleware ✅ 
     * Endpoint ✅ 
* Tutorials
     * Your First ZIO HTTP App ✅ 
     * Deploying a ZIO HTTP App
     * Testing Your ZIO HTTP App
* How-to-guides
     * (This should be based off the list of topics at http4k, but should also include ZIO HTTP use cases around Endpoint, middleware, etc.) ✅ 
* Reference
     *  API Docs
     * Server backend
     * Websockets
     * JSON Handling 
     * Metrics
     * Request Logging
     * (etc.)
* Performance ✅ 
* FAQ ✅ 

/claim #2198

### CODE REVIEW

**Comment by khajavi** [comment]
To avoid conflict, I'm informing you that I'm going to work on this PR to make it ready for merge. Please let me know if you have any notes. @daveads

**Comment by khajavi** [comment]
@daveads 
I have fixed the CI and now it is showing green. I noticed that you wrote some good code, but it seems like you haven't worked on it in a month. Can you please let me know what your plan is? Do you intend to continue working on it or would you like me to assist you in completing the PR and splitting the bouncy?

File: README.md [hunk 1]
```diff
@@ -13,45 +13,170 @@ ZIO HTTP is a scala library for building http apps. It is powered by ZIO and [Ne
...
-- Newer library versions, starting in 2023 and resulting from the ZIO organization (`dev.zio`) started with `0.0.x`, reaching `1.0.0` release candidates in April of 2023
+- Typesafe HTTP contract definition with support for path parameters, query parameters, headers, and request/response bodies
+- Automatic validation of incoming requests based on contract definition
+- Self-documenting routes with built-in support for OpenAPI (Swagger) descriptions
 
-## Getting Started
+Templating:
 
-A simple Http server can be built using a few lines of code.
+- Pluggable templating system support for popular template engines such as Dust, Freemarker, Handlebars, and Thymeleaf
```
Comment (khajavi) [inline README.md:98]: I see a lot of tools and technologies mentioned in this file that I doubt we currently support. Please make sure that they are related to ZIO‌ HTTP.

Comment (daveads) [reply README.md:98]: @khajavi yea i intend on working on it  this new week.

Comment (khajavi) [reply README.md:98]: Great to hear that!

File: docs/concepts/request-handling.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+In ZIO HTTP, request handling is the process of receiving an HTTP request, processing it, and generating an appropriate HTTP response. ZIO HTTP provides a functional and composable approach to request handling using the HttpApp type.
+
+The `HttpApp` type represents an HTTP application, which is essentially a function that takes an incoming Request and returns a Response. It is defined as type `HttpApp[R, E] = Request => ZIO[R, E, Response[E]]`, where `R` represents the required environment, `E` represents the potential error type, and `Response[E]` represents the response type.
+
+Here's an overview of the request handling process in ZIO HTTP:
+
+- **Defining an `HttpApp`**: To handle requests, you define one or more `HttpApp` values. Each `HttpApp` is responsible for processing a specific type of request or a specific path pattern.
+
+- **Request Matching**: The `Http.collect` or `Http.collectM` combinator is used to pattern match on the incoming `Request`. It allows you to define matching criteria based on HTTP method, path, headers, etc.
```
Comment (khajavi) [inline docs/concepts/request-handling.md]: Please note that the functions 'collect' and 'collectM' are no longer available and have been deprecated.

File: docs/index.md [hunk 3]
```diff
@@ -13,42 +13,167 @@ ZIO HTTP is a scala library for building http apps. It is powered by ZIO and [Ne
...
+Serverless:
+
+- Function-based support for building serverless HTTP and event-driven applications
+- Adapters for AWS Lambda, Google Cloud Functions, Azure Functions, and other serverless platforms
+- Custom AWS Lambda runtime for improved performance and reduced startup time
+
+Contract:
 
-- Older library versions `1.x` or `2.x` with organization `io.d11` of ZIO Http are derived from Dream11, the organization that donated ZIO Http to the ZIO organization in 2022. 
-- Newer library versions, starting in 2023 and resulting from the ZIO organization (`dev.zio`) started with `0.0.x`, reaching `1.0.0` release candidates in April of 2023
```
Comment (khajavi) [inline docs/index.md:22]: This section is important and I believe it should remain in the documentation.

File: docs/tutorials/deploying-a-zio-http-app.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
+---
+id: deploying-zio-http-application
+title: Deploying Zio-http Application 
+---
```
Comment (khajavi) [inline docs/tutorials/deploying-a-zio-http-app.md]: If you really wish (although I don't recommend it) to relocate this article from [zio.dev](https://zio.dev/guides/tutorials/deploy-a-zio-application-using-docker), kindly open a separate PR on [github.com/zio/zio](https://github.com/zio/zio) and eliminate it from zio.dev.

File: docs/tutorials/testing-your-zio-http-app.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
+---
+id: testing-your-zio-http-app
+title: Testing zio http app
+---
```
Comment (khajavi) [inline docs/tutorials/testing-your-zio-http-app.md:4]: This article must be revised as it is irrelevant to the ZIO HTTP project.

**Comment by khajavi** [comment]
Hello @daveads
I added a couple of other comments to your work. So I believe this is still far from being completed.
Also before asking for another review, please make sure all the examples are compile-safe, using `mdoc` modifier.
Anyway, please let me know if you're going to continue working on this. I'm here to help if you want.

**Comment by 987Nabil** [comment]
In relation to @khajavi I converted this to a draft

**Comment by daveads** [comment]
> In relation to @khajavi I converted this to a draft

Alright

File: docs/tutorials/deploying-a-zio-http-app.md (hunk 4 above)
Comment (daveads) [reply docs/tutorials/deploying-a-zio-http-app.md]: will remove it

**Comment by jdegoes** [comment]
@daveads Please re-open if you get a chance to complete!

**Pull request closed without merging**